api.go

```diff
@@ -544,10 +551,15 @@ func (c Client) executeMethod(method string, metadata requestMetadata) (res *htt
 
 		// For errors verify if its retryable otherwise fail quickly.
 		errResponse := ToErrorResponse(httpRespToErrorResponse(res, metadata.bucketName, metadata.objectName))
+
+		// Save the body.
+		errBodySeeker = bytes.NewReader(errBodyBytes)
+		res.Body = ioutil.NopCloser(errBodySeeker)
+
 		// Bucket region if set in error response and the error code dictates invalid region,
 		// we can retry the request with the new region.
-		if errResponse.Code == "InvalidRegion" && errResponse.Region != "" {
-			c.bucketLocCache.Set(metadata.bucketName, errResponse.Region)
+		if errResponse.Code == "InvalidRegion" && errResponse.Headers.Get("x-amz-bucket-region") != "" {
```

Comment:
Don't do this, set the Region properly and remove Headers from errorResponse structure. Let me send a PR to this PR.